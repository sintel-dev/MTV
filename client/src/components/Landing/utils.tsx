export const colorSchemes = {
  tag: [
    '#FFCD00', // investigate
    '#7CA8FF', // do not investigate
    '#A042FF', // postpone
    '#F64242', // problem
    '#F5FF00', // previously seen
    '#45F642', // normal
    '#C7C7C7', // untagged
  ],
};

// @TODO - later investigation for refactoring (fromIDToTag/fromTagToId)
export const tagSeq = [
  'Investigate',
  'Do not Investigate',
  'Postpone',
  'Problem',
  'Previously seen',
  'Normal',
  'Untagged',
];
const tagClassNames = ['investigate', 'not_investigate', 'postpone', 'problem', 'seen', 'normal', 'untagged'];

export const fromTagToClassName = (tag: string): string => {
  switch (tag) {
    case tagSeq[0]:
      return tagClassNames[0];
    case tagSeq[1]:
      return tagClassNames[1];
    case tagSeq[2]:
      return tagClassNames[2];
    case tagSeq[3]:
      return tagClassNames[3];
    case tagSeq[4]:
      return tagClassNames[4];
    case tagSeq[5]:
      return tagClassNames[5];
    default:
      return 'untagged';
  }
};

export const fromIDtoTag = (id: string): string => {
  switch (id) {
    case '1':
      return tagSeq[0];
    case '2':
      return tagSeq[1];
    case '3':
      return tagSeq[2];
    case '4':
      return tagSeq[3];
    case '5':
      return tagSeq[4];
    case '6':
      return tagSeq[5];
    default:
      return 'untagged';
  }
};

export const fromTagToID = (tag: string): string => {
  switch (tag) {
    case tagSeq[0]:
      return '1';
    case tagSeq[1]:
      return '2';
    case tagSeq[2]:
      return '3';
    case tagSeq[3]:
      return '4';
    case tagSeq[4]:
      return '5';
    case tagSeq[5]:
      return '6';
    default:
      return 'untagged';
  }
};

export const getTagColor = (tag: string): string => {
  let colorIdx: number | undefined;
  for (let i = 0; i < tagSeq.length; i += 1) {
    if (tagSeq[i] === tag) {
      colorIdx = i;
    }
  }
  if (colorIdx === undefined) {
    colorIdx = 6;
  }
  return colorSchemes.tag[colorIdx];
};
