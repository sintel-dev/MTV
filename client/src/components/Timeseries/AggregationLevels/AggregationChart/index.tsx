import React, { Component } from 'react';
import Loader from 'src/components/Common/Loader';
import { connect } from 'react-redux';
import * as d3 from 'd3';
import { RootState } from 'src/model/types';
import { getDatarunDetails, getCurrentEventDetails } from 'src/model/selectors/datarun';
import { getAggregationTimeLevel } from 'src/model/selectors/aggregationLevels';
import { FocusChartConstants } from '../../FocusChart/Constants';
import { getAggregationChartCoords, timestampToDate } from './Utils';

const { MIN_VALUE, MAX_VALUE, TRANSLATE_LEFT, CHART_MARGIN } = FocusChartConstants;

type State = {
  width: number;
  height: number;
  zoomValue: any;
};
type StateProps = ReturnType<typeof mapState>;

class AggregationChart extends Component<StateProps, State> {
  constructor(props) {
    super(props);
    this.state = {
      width: 0,
      height: 450,
      zoomValue: 1,
    };
  }

  componentDidMount() {
    const { width } = getAggregationChartCoords();
    this.setState({
      width,
    });
  }

  componentDidUpdate() {
    this.renderChartAxis();
    const { width } = getAggregationChartCoords();
    if (this.state.width !== width) {
      this.setState(
        {
          width,
        },
        () => this.initZoom(),
      );
    }
  }

  parseDataInterval(data) {
    const { currentAggregationLevel } = this.props;

    let initialStartDate = data[0];
    let initialStopDate = new Date(initialStartDate[0] + currentAggregationLevel.timeInMiliseconds).getTime();
    let processedInterval = [initialStartDate];

    data.forEach((currentData) => {
      if (currentData[0] >= initialStopDate) {
        processedInterval.push(currentData);
        initialStopDate = new Date(currentData[0] + currentAggregationLevel.timeInMiliseconds).getTime();
      }
    });
    processedInterval.push(data[data.length - 1]);
    return processedInterval;
  }

  drawLine(data) {
    const { xCoord, yCoord } = this.getScale();
    const { zoomValue } = this.state;
    const xCoordCopy = xCoord.copy();
    const { start_time, stop_time } = this.props.currentEventDetails;
    const filteredData = data.filter((currentData) => currentData[0] >= start_time && currentData[0] <= stop_time);

    if (zoomValue !== 1) {
      xCoord.domain(zoomValue.rescaleX(xCoordCopy).domain());
    }

    const line = d3
      .line()
      .x((d) => xCoord(d[0]))
      .y((d) => yCoord(d[1]));
    return line(this.parseDataInterval(filteredData));
  }

  getScale() {
    const { width, height } = this.state;
    const { dataRun, currentEventDetails } = this.props;
    const { start_time, stop_time } = currentEventDetails;
    const { maxTimeSeries } = dataRun;
    const [minTY, maxTY] = d3.extent(maxTimeSeries, (time: Array<number>) => time[1]);
    const drawableWidth = width - TRANSLATE_LEFT;
    const drawableHeight = height - 3.5 * CHART_MARGIN;
    const xCoord = d3.scaleTime().range([0, drawableWidth]);
    const yCoord = d3.scaleLinear().range([drawableHeight, 0]);

    const minX = Math.min(MIN_VALUE, start_time);
    const maxX = Math.max(MAX_VALUE, stop_time);

    const minY = Math.min(MIN_VALUE, minTY);
    const maxY = Math.max(MAX_VALUE, maxTY);

    xCoord.domain([minX, maxX]);
    yCoord.domain([minY, maxY]);

    return { xCoord, yCoord };
  }

  initZoom() {
    const { width, height } = this.state;
    const zoomWidth = width - TRANSLATE_LEFT;

    const zoom = d3
      .zoom()
      .scaleExtent([1, zoomWidth])
      .translateExtent([
        [0, 0],
        [zoomWidth, height],
      ])
      .extent([
        [0, 0],
        [zoomWidth, height],
      ])
      .on('zoom', () => zoomHandler());

    d3.select('.aggregation-zoom').call(zoom);

    const zoomHandler = () => {
      let currentZoom = d3.event.transform;

      if (currentZoom === 1) {
        return;
      }
      this.setState({
        zoomValue: currentZoom,
      });
    };
  }

  drawChartData() {
    const { width, height } = this.state;
    const { timeSeries } = this.props.dataRun;

    return (
      width > 0 && (
        <g className="aggregation-focus" transform={`translate(${TRANSLATE_LEFT - 10}, ${CHART_MARGIN})`}>
          <defs>
            <clipPath id="aggregationClip">
              <rect width={width - TRANSLATE_LEFT} height={height} />
            </clipPath>
          </defs>
          <g className="aggregation-data" clipPath="url(#aggregationClip)">
            <path className="aggregation-wawes" d={this.drawLine(timeSeries)} />
            <rect className="aggregation-zoom" width={width} height={height} />
          </g>
          <g className="aggregation-axis">
            <g className="axis axis--x" transform={`translate(0, ${height - 3.5 * CHART_MARGIN})`} />
            <g className="axis axis--y" />
          </g>
        </g>
      )
    );
  }

  renderChartAxis() {
    const { xCoord, yCoord } = this.getScale();
    const xCoordCopy = xCoord.copy();
    const { zoomValue } = this.state;

    if (zoomValue !== 1) {
      xCoord.domain(zoomValue.rescaleX(xCoordCopy).domain());
    }

    const xAxis = d3.axisBottom(xCoord);
    const yAxis = d3.axisLeft(yCoord);

    d3.select('.aggregation-axis .axis.axis--x').call(xAxis);
    d3.select('.aggregation-axis .axis.axis--y').call(yAxis).call(yAxis.ticks(5, ',f'));
  }

  renderEventDetails() {
    const { currentEventDetails } = this.props;
    const { start_time, stop_time, signal } = currentEventDetails;
    return (
      <div className="event-details">
        <ul>
          <li>
            <p>Start</p>
            <p>{timestampToDate(start_time)}</p>
          </li>
          <li>
            <p>Stop</p>
            <p>{timestampToDate(stop_time)}</p>
          </li>
          <li>
            <p>Dataset</p>
            <p>{signal}</p>
          </li>
          {/* Addiitonal info required */}
          {/* <li>
            <p>Experiment</p>
            <p>Will Follow Soon</p>
          </li> */}
        </ul>
      </div>
    );
  }

  render() {
    const { width, height } = this.state;
    return (
      width > 0 && (
        <Loader isLoading={width === 0}>
          {this.renderEventDetails()}
          <div className="aggregation-chart">
            <svg width={width} height={height} id="aggregationChart">
              {this.drawChartData()}
            </svg>
          </div>
        </Loader>
      )
    );
  }
}

const mapState = (state) => ({
  dataRun: getDatarunDetails(state),
  currentEventDetails: getCurrentEventDetails(state),
  currentAggregationLevel: getAggregationTimeLevel(state),
});

export default connect<StateProps, {}, {}, RootState>(mapState)(AggregationChart);
