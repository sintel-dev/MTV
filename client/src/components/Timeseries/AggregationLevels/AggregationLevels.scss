.modal{
    &.aggregation-levels-modal{
        .close{
            color: #fff;
        }

        .modal-dialog{
            width: 1080px;
            max-width: 90%;
            transform: translate(0, 0) !important;
        }

        .modal-content {
            border-radius: 13px;
            border: 1px solid #60666C;
            .modal-body{
                overflow: hidden;
            }
            .aggregation-chart {
                background: #2C3034;
                border-bottom-left-radius: 12px;
                border-bottom-right-radius: 12px;;
            }
        }

        .aggregation-wrapper {
            margin-bottom: 30px;

            .headings{
                display: flex;
                margin-bottom: 35px;

                h4{
                    align-items: flex-start;
                    font-size: 16px;
                    margin: 0;
                    padding: 0;
                    flex: 1;
                }

                .note{
                    align-items: flex-end;
                    color: #A0A0A0
                }
            }

            .aggregation-controls{
                display: flex;
                justify-content: space-between;

                button{
                    height: 34px;
                    border-radius: 5px;
                    background: #2D3037;
                    border: 2px solid #2D3037;
                    color: #fff;
                    cursor: pointer;
                    transition: border-color .2s ease-in-out;

                    &.active {
                        border-color: #A2AAB2;
                    }
                    &:hover{
                        border-color: #A0AAB3;
                    }
                }
            }
        }

        .aggregation-axis{
            font-size: 10px;
            color: #8A9097;
            .tick{
                line{
                    stroke: #fff;
                }
            }
        }

        .event-details{
            background: #26282E;
            border-radius: 4px;

            padding: 20px 30px;
            box-sizing: border-box;

            ul{
                display: flex;
                justify-content: space-between;
            }
        }

        .aggregation-data {
            path{
                &.aggregation-wawes{
                    fill: transparent;
                    stroke: #2474f4;
                }
            }
            .aggregation-zoom{
                fill: transparent;
                cursor: move;
            }
        }
    }
}