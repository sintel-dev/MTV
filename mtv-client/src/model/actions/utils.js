const baseUrl = '/api/v1/';

export const api = {
    get(url, dataType = 'json') {
        const apiUrl = `${baseUrl}${url}`;
        const promise = fetch(apiUrl, {
            mode: 'cors',
            headers: {
                'Access-Control-Allow-Origin': '*',
            },
        });
        return promise.then(response => {
            if (response.status !== 204) {
                return dataType !== 'json' ? response.text : response.json();
            }
            return null;
        })
        .catch(err => console.error(err));
    },

    // post(url, data) {
    //     const apiUrl = `${baseUrl}${url}`;
    //     return fetch(apiUrl, {
    //         mode: 'cors',
    //         method: 'POST',
    //         headers: {
    //             'Access-Control-Allow-Origin': '*',
    //             'Content-Type': 'application/json',
    //             // 'Authorization': lscache.get('authorization')
    //         },
    //         body: JSON.stringify(data)
    //     });
    // }
};
