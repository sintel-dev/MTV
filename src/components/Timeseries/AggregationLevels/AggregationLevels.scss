.modal {
    &.aggregation-levels-modal {
        .close {
            color: #fff;
        }

        .modal-dialog {
            width: 1080px;
            max-width: 90%;
            transform: translate(0, 0) !important;
        }

        .modal-content {
            border-radius: 13px;
            border: 1px solid #60666c;
            .modal-body {
                overflow: hidden;
            }
            .aggregation-chart {
                background: #2c3034;
                border-bottom-left-radius: 12px;
                border-bottom-right-radius: 12px;
            }
        }

        .aggregation-axis {
            font-size: 10px;
            color: #8a9097;
            .tick {
                line {
                    stroke: #fff;
                }
            }
        }

        .event-details {
            background: #26282e;
            border-radius: 4px;

            padding: 20px 30px;
            box-sizing: border-box;

            ul {
                display: flex;
                justify-content: space-between;
            }
        }

        .aggregation-data {
            path {
                &.aggregation-wawes {
                    fill: transparent;
                    stroke: #2474f4;
                }
            }
            .aggregation-zoom {
                fill: transparent;
                cursor: move;
            }
        }
    }
}

.aggregation-chart {
    padding: 16px 0 0 0;
    position: relative;
    .aggregation-data {
        path {
            &.aggregation-event,
            &.aggregation-wawes {
                stroke-width: 2px;
                fill: transparent;
            }

            &.aggregation-wawes {
                stroke: rgba(#b2c1ff, 0.6);
            }
        }

        .event-highight {
            rect.event-area {
                fill: rgba(#5a627c, 0.4);
            }
        }
    }
    .aggregation-focus {
        .aggregation-zoom {
            fill: transparent;
            cursor: move;
        }
    }

    .aggregation-wrapper {
        margin: 0 16px;
        // margin-bottom: 30px;
        // margin: 16px;

        .aggregation-controls {
            display: flex;
            justify-content: space-between;

            button {
                background: #23242b;
                border-radius: 16px;
                // height: 34px;
                // border-radius: 5px;
                // background: #2d3037;
                // border: 2px solid #2d3037;
                color: #fff;
                cursor: pointer;
                border: none;
                padding: 8px 16px;

                &.active,
                &:hover {
                    color: #23242b;
                    background: #fff;
                }
            }
        }
        .context-info {
            span {
                display: inline-block;
                margin-right: 16px;
            }
            .tag-select {
                .tag-options__menu {
                    min-height: 30px;
                    font-size: 14px;
                }
                .tag-options__control {
                    min-height: 30px;
                }
                .tag-options__menu,
                .tag-options__control {
                    min-width: 65px;
                }
            }
        }
    }
}
