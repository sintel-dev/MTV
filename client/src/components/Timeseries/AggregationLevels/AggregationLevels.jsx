import React, { Component } from 'react';
import { connect } from 'react-redux';
import * as d3 from 'd3';
import { getCurrentEventDetails, getDatarunDetails, getCurrentChartStyle } from 'src/model/selectors/datarun';
import {
  getAggregationTimeLevel,
  getSignalRawData,
  getIsSigRawLoading,
  getEventInterval,
  getAggregationZoomValue,
} from 'src/model/selectors/aggregationLevels';
import {
  setAggregationLevelAction,
  getSignalRawDataAction,
  setContextValueAction,
  updateAggregationZoomAction,
} from 'src/model/actions/aggregationLevels';
import Loader from 'src/components/Common/Loader';
import Dropdown from 'src/components/Common/Dropdown';
import { FocusChartConstants, colorSchemes } from '../FocusChart/Constants';
import { timeIntervals } from './AggregationChart/Utils';
import './AggregationLevels.scss';

const { MIN_VALUE, MAX_VALUE, TRANSLATE_LEFT, CHART_MARGIN, TIME_INTERVALS_HEIGHT } = FocusChartConstants;
const TRANSLATE_CHART = TRANSLATE_LEFT + 20;

class AggregationLevels extends Component {
  componentDidMount() {
    this.props.getSignalRawData();
    this.props.toggleTooltip();
  }

  componentDidUpdate(prevProps) {
    this.initZoom();

    if (prevProps.signalRawData !== this.props.signalRawData && this.props.signalRawData !== null) {
      this.updateAxisOnZoom();
    }
  }

  getEventInterval() {
    const { dataRun, currentEventDetails } = this.props;
    const { start_time, stop_time } = currentEventDetails;
    const { timeSeries } = dataRun;
    const eventInterval = timeSeries.filter((current) => current[0] >= start_time && current[0] <= stop_time);

    const startIndex = timeSeries.findIndex((element) => start_time - element[0] < 0) - 1 - eventInterval.length;
    const stopIndex = timeSeries.findIndex((element) => stop_time - element[0] < 0) + eventInterval.length;

    const eventWrapper = timeSeries.slice(startIndex, stopIndex);
    return { eventInterval, eventWrapper, start_time, stop_time, startIndex, stopIndex };
  }

  initZoom() {
    const { width, height } = this.props;
    const zoomWidth = width - TRANSLATE_CHART - 2 * CHART_MARGIN;
    const zoom = d3
      .zoom()
      .scaleExtent([1, Infinity])
      .translateExtent([
        [0, 0],
        [zoomWidth, height],
      ])
      .extent([
        [0, 0],
        [zoomWidth, height],
      ])
      .on('zoom', () => zoomHandler());

    d3.select('.aggregation-zoom').call(zoom);

    const zoomHandler = () => {
      let currentZoom = d3.event.transform;
      if (currentZoom === 1) {
        return;
      }
      this.props.updateZoom(currentZoom);
    };
  }

  getScale() {
    const { width, height, signalRawData } = this.props;
    const chartWidth = width - TRANSLATE_CHART - 2 * CHART_MARGIN;
    const [minTY, maxTY] = d3.extent(signalRawData, (time) => time[1]);
    const chartHeight = height - TIME_INTERVALS_HEIGHT;
    const drawableHeight = chartHeight - 3.5 * CHART_MARGIN;

    const xCoord = d3.scaleTime().range([0, chartWidth]);
    const yCoord = d3.scaleLinear().range([drawableHeight, 0]);

    const start = signalRawData[0][0];
    const stop = signalRawData[signalRawData.length - 1][0];

    const minX = Math.min(MIN_VALUE, start);
    const maxX = Math.max(MAX_VALUE, stop);

    const minY = Math.min(MIN_VALUE, minTY);
    const maxY = Math.max(MAX_VALUE, maxTY);

    xCoord.domain([minX, maxX]);
    yCoord.domain([minY, maxY]);

    return { xCoord, yCoord };
  }

  drawLine(eventInterval) {
    const { zoomValue } = this.props;
    const { currentChartStyle } = this.props;
    const { xCoord, yCoord } = this.getScale();
    const xCoordCopy = xCoord.copy();

    if (zoomValue !== 1) {
      xCoord.domain(zoomValue.rescaleX(xCoordCopy).domain());
    }

    const line = d3
      .line()
      .x((d) => xCoord(d[0]))
      .y((d) => yCoord(d[1]));
    line.curve(currentChartStyle === 'linear' ? d3.curveLinear : d3.curveStepBefore);

    return line(eventInterval);
  }

  updateAxisOnZoom() {
    const { zoomValue } = this.props;
    const { xCoord, yCoord } = this.getScale();
    const xCoordCopy = xCoord.copy();

    if (zoomValue !== 1) {
      xCoord.domain(zoomValue.rescaleX(xCoordCopy).domain());
    }

    const xAxis = d3.axisBottom(xCoord);
    const yAxis = d3.axisLeft(yCoord);

    d3.select('.axis.axis--x').call(xAxis);
    d3.select('.axis.axis--y')
      .call(yAxis)
      .call(yAxis.ticks(5, ',f').tickFormat(d3.format('.4s')));
  }

  renderEventArea() {
    const { zoomValue } = this.props;
    const { height, currentEventDetails, eventInterval } = this.props;
    const { xCoord } = this.getScale();
    const chartHeight = height - TIME_INTERVALS_HEIGHT;

    const xCoordCopy = xCoord.copy();

    if (zoomValue !== 1) {
      xCoord.domain(zoomValue.rescaleX(xCoordCopy).domain());
    }

    const eventTag = colorSchemes[currentEventDetails.tag] || colorSchemes.Untagged;

    const eventWidth = Math.max(xCoord(eventInterval[eventInterval.length - 1][0]) - xCoord(eventInterval[0][0]));
    const eventHeight = chartHeight - 3.5 * CHART_MARGIN;
    const translateEvent = xCoord(eventInterval[0][0]);

    return (
      <g className="event-highight">
        <rect className="event-tag" width={eventWidth} height="10" y={0} x={translateEvent} fill={eventTag} />
        <rect className="event-area" width={eventWidth} height={eventHeight} y={0} x={translateEvent} />
      </g>
    );
  }

  drawChartData() {
    const { signalRawData, eventInterval } = this.props;
    const { width, height, currentEventDetails, isSignalRawLoading } = this.props;
    const chartWidth = width - TRANSLATE_CHART - 2 * CHART_MARGIN;
    const chartHeight = height - TIME_INTERVALS_HEIGHT; // time granulation height;
    const pathClassName = currentEventDetails.tag?.replace(/\s/g, '_').toLowerCase() || 'untagged';
    if (isSignalRawLoading) {
      return null;
    }
    return (
      width > 0 && (
        <g className="aggregation-focus" transform={`translate(${TRANSLATE_CHART}, ${CHART_MARGIN})`}>
          <defs>
            <clipPath id="aggregationClip">
              <rect width={chartWidth} height={chartHeight} />
            </clipPath>
          </defs>
          <g className="aggregation-data" clipPath="url(#aggregationClip)">
            <path className="aggregation-wawes" d={this.drawLine(signalRawData, 'eventWrapper')} />
            <path
              className={`evt-highlight aggregation-event ${pathClassName}`}
              d={this.drawLine(eventInterval, 'eventInterval')}
            />

            {this.renderEventArea()}
          </g>
          <g className="aggregation-axis">
            <g className="axis axis--x" transform={`translate(0, ${chartHeight - 3.5 * CHART_MARGIN})`} />
            <g className="axis axis--y" />
            {this.updateAxisOnZoom()}
          </g>
          <rect className="aggregation-zoom" width={chartWidth} height={chartHeight} />
        </g>
      )
    );
  }

  renderIntervalLevels() {
    const { setAggregationLevel, currentAggregationLevel, setContextInfo, updateZoom } = this.props;
    const contextInfoValues = [
      { value: 1, label: '1x' },
      { value: 2, label: '2x' },
      { value: 3, label: '3x' },
    ];

    const dropDownProps = {
      isMulti: false,
      closeMenuOnSelect: true,
      onChange: (event) => {
        setContextInfo(event.value);
        updateZoom(1);
      },
      placeholder: '1x',
      options: contextInfoValues,
      formatLabel: false,
    };
    return (
      <div className="aggregation-wrapper">
        <ul className="aggregation-controls">
          {timeIntervals.map((level) => (
            <li key={level}>
              <button
                type="button"
                onClick={() => setAggregationLevel(level)}
                className={currentAggregationLevel.selectedLevel === level ? 'active' : ''}
              >
                {level}
              </button>
            </li>
          ))}
          <li className="context-info">
            <span>Context info</span>
            <Dropdown {...dropDownProps} />
          </li>
        </ul>
      </div>
    );
  }

  render() {
    const { width, height, isSignalRawLoading } = this.props;

    return (
      <div className="aggregation-chart" id="aggregationChart" style={{ position: 'relative', width, height }}>
        {this.renderIntervalLevels()}
        <Loader isLoading={isSignalRawLoading}>
          <svg width={width} height={height}>
            {this.drawChartData()}
          </svg>
        </Loader>
      </div>
    );
  }
}

export default connect(
  (state) => ({
    currentEventDetails: getCurrentEventDetails(state),
    dataRun: getDatarunDetails(state),
    currentChartStyle: getCurrentChartStyle(state),
    currentAggregationLevel: getAggregationTimeLevel(state),
    isSignalRawLoading: getIsSigRawLoading(state),
    signalRawData: getSignalRawData(state),
    eventInterval: getEventInterval(state),
    zoomValue: getAggregationZoomValue(state),
  }),
  (dispatch) => ({
    setAggregationLevel: (periodLevel) => dispatch(setAggregationLevelAction(periodLevel)),
    getSignalRawData: () => dispatch(getSignalRawDataAction()),
    setContextInfo: (contextValue) => dispatch(setContextValueAction(contextValue)),
    updateZoom: (zoomValue) => dispatch(updateAggregationZoomAction(zoomValue)),
  }),
)(AggregationLevels);
