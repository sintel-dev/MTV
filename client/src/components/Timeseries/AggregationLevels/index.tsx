import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Modal, ModalBody } from 'react-bootstrap';
import { RootState } from 'src/model/types';
import { toggleAggregationModal, setAggregationLevelAction } from '../../../model/actions/aggregationLevels';
import { getIsAggregationModalOpen, getAggregationTimeLevel } from '../../../model/selectors/aggregationLevels';
import AggregationChart from './AggregationChart';
import { timeIntervals } from './AggregationChart/Utils';
import './AggregationLevels.scss';

type StateProps = ReturnType<typeof mapState>;
type DispatchProps = ReturnType<typeof mapDispatch>;
type Props = StateProps & DispatchProps;

class AggregationLevels extends Component<Props, {}> {
  renderLevelIntervals() {
    const { setAggregationLevel, currentAggregationLevel } = this.props;

    return (
      <div className="aggregation-wrapper">
        <div className="headings">
          <h4>Aggregation levels:</h4>
          <span className="note">
            <strong>Note:</strong> Fine aggregation levels will require more time to load.
          </span>
        </div>
        <ul className="aggregation-controls">
          {timeIntervals.map((level) => (
            <li key={level}>
              <button
                type="button"
                onClick={() => setAggregationLevel(level)}
                className={currentAggregationLevel.selectedLevel === level ? 'active' : ''}
              >
                {level}
              </button>
            </li>
          ))}
        </ul>
      </div>
    );
  }

  render() {
    const { isModalOpen, toggleModalState } = this.props;
    return (
      <Modal show={isModalOpen} centered className="aggregation-levels-modal" onHide={() => toggleModalState(false)}>
        <Modal.Header closeButton>Signal Aggregation Levels</Modal.Header>
        <ModalBody>
          {this.renderLevelIntervals()}
          <AggregationChart />
        </ModalBody>
      </Modal>
    );
  }
}

const mapState = (state: RootState) => ({
  isModalOpen: getIsAggregationModalOpen(state),
  currentAggregationLevel: getAggregationTimeLevel(state),
});

const mapDispatch = (dispatch: Function) => ({
  toggleModalState: (modalState) => dispatch(toggleAggregationModal(modalState)),
  setAggregationLevel: (periodLevel) => dispatch(setAggregationLevelAction(periodLevel)),
});

export default connect<StateProps, DispatchProps, {}, RootState>(mapState, mapDispatch)(AggregationLevels);
