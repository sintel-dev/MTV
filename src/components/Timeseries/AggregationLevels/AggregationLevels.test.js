import React from 'react';
import { TestWrapper, renderWithStore } from 'src/tests/utils';
import { AggregationLevels } from './AggregationLevels';

describe('Testing Aggregation levels component', () => {
  const aggregationProps = {
    currentAggregationLevel: {
      selectedLevel: '30 hours',
      timeInMiliseconds: 108000000,
    },
    currentEventDetails: {
      tag: 'Untagged',
    },
    width: 200,
    height: 700,
    signalRawData: [
      [1536451200000, 1945.5399780273438],
      [1536559200000, 1987.1500244140625],
      [1536667200000, 1990],
      [1536775200000, 1989.8699951171875],
      [1536883200000, 1970.18994140625],
      [1536991200000, 1970.18994140625],
      [1537099200000, 1908.030029296875],
      [1537207200000, 1941.050048828125],
    ],
    zoomValue: 1,
    currentChartStyle: 'linear',
    eventInterval: [
      [1550966400000, 173.5999984741211],
      [1551074400000, 174.3300018310547],
      [1551182400000, 174.8699951171875],
      [1551290400000, 173.14999389648438],
      [1551398400000, 174.97000122070312],
      [1551506400000, 174.97000122070312],
      [1551614400000, 175.85000610351562],
      [1551722400000, 175.52999877929688],
      [1551830400000, 173.51000213623047],
      [1551938400000, 172.91000366210938],
    ],
  };
  const AggregationComponent = renderWithStore(
    {},
    <TestWrapper>
      <AggregationLevels {...aggregationProps} />
    </TestWrapper>,
  );

  it('Should render without crashing', () => {
    expect(AggregationComponent).toMatchSnapshot();
  });
});
