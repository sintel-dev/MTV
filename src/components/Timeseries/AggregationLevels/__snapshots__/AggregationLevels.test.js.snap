// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Aggregation levels component Should render without crashing 1`] = `
<div
  class="aggregation-chart"
  id="aggregationChart"
  style="position:relative;width:200px;height:700px"
>
  <div
    class="aggregation-wrapper"
  >
    <ul
      class="aggregation-controls"
    >
      <li>
        <button
          class=""
          type="button"
        >
          30 days
        </button>
      </li>
      <li>
        <button
          class=""
          type="button"
        >
          21 days
        </button>
      </li>
      <li>
        <button
          class=""
          type="button"
        >
          14 days
        </button>
      </li>
      <li>
        <button
          class=""
          type="button"
        >
          7 days
        </button>
      </li>
      <li>
        <button
          class=""
          type="button"
        >
          5 days
        </button>
      </li>
      <li>
        <button
          class=""
          type="button"
        >
          3 days
        </button>
      </li>
      <li>
        <button
          class=""
          type="button"
        >
          2 days
        </button>
      </li>
      <li>
        <button
          class=""
          type="button"
        >
          1 day
        </button>
      </li>
      <li
        class="context-info"
      >
        <span>
          Context info
        </span>
        <div
          class="tag-select css-2b097c-container"
        >
          <div
            class="tag-options__control css-yk16xz-control"
          >
            <div
              class="tag-options__value-container css-g1d714-ValueContainer"
            >
              <div
                class="tag-options__placeholder css-1wa3eu0-placeholder"
              >
                1x
              </div>
              <input
                aria-autocomplete="list"
                class="css-62g3xt-dummyInput"
                id="react-select-2-input"
                readonly=""
                tabindex="0"
                value=""
              />
            </div>
            <div
              class="tag-options__indicators css-1hb7zxy-IndicatorsContainer"
            >
              <span
                class="tag-options__indicator-separator css-1okebmr-indicatorSeparator"
              />
              <div
                aria-hidden="true"
                class="tag-options__indicator tag-options__dropdown-indicator css-tlfecz-indicatorContainer"
              >
                <svg
                  aria-hidden="true"
                  class="css-6q0nyr-Svg"
                  focusable="false"
                  height="20"
                  viewBox="0 0 20 20"
                  width="20"
                >
                  <path
                    d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </li>
    </ul>
  </div>
  <svg
    height="700"
    width="200"
  >
    <g
      class="aggregation-focus"
      transform="translate(85, 10)"
    >
      <defs>
        <clipPath
          id="aggregationClip"
        >
          <rect
            height="635"
            width="95"
          />
        </clipPath>
      </defs>
      <g
        class="grid-lines"
        id="aggGridLines"
      />
      <g
        class="aggregation-data"
        clip-path="url(#aggregationClip)"
      >
        <path
          class="aggregation-wawes"
          d="M0,325.4364123199175L13.571428571428571,20.861119467221023L27.142857142857142,0L40.714285714285715,0.9516037277847333L54.285714285714285,145.0047505726004L67.85714285714286,145.0047505726004L81.42857142857143,600L95,358.3015883889453"
        />
        <path
          class="evt-highlight aggregation-event untagged"
          d="M1824,13295.600712883732L1837.5714285714284,13290.257268078236L1851.1428571428569,13286.304650200002L1864.7142857142858,13298.89463557598L1878.2857142857144,13285.572630149307L1891.857142857143,13285.572630149307L1905.4285714285713,13279.131211113065L1919,13281.473585922819L1932.5714285714284,13296.259463915008L1946.1428571428569,13300.651304992718"
        />
        <g
          class="event-highight"
        >
          <rect
            class="event-tag"
            fill="#D5D5D5"
            height="10"
            width="122.14285714285688"
            x="1824"
            y="0"
          />
          <rect
            class="event-area"
            height="600"
            width="122.14285714285688"
            x="1824"
            y="0"
          />
        </g>
      </g>
      <g
        class="aggregation-axis"
      >
        <g
          class="axis axis--x"
          transform="translate(0, 600)"
        />
        <g
          class="axis axis--y"
        />
      </g>
      <rect
        class="aggregation-zoom"
        height="635"
        width="95"
      />
    </g>
  </svg>
</div>
`;
